pool.go

```diff
@@ -13,17 +15,66 @@ func newBufPool(depth, bufLen int) *bufPool {
 }
 
 func (p *bufPool) Get() []byte {
-	select {
-	case b := <-p.ch:
-		return b
-	default:
+	if p == nil {
+		// functional default: no reuse.
 		return make([]byte, p.blen)
 	}
+
+	for {
+		select {
+		case b := <-p.ch:
+			if cap(b) < p.blen {
+				// just in case: throw away any buffer with insufficient capacity.
+				continue
+			}
+
+			return b[:p.blen]
+
+		default:
+			return make([]byte, p.blen)
+		}
+	}
 }
 
 func (p *bufPool) Put(b []byte) {
+	if p == nil {
+		// functional default: no reuse.
+		return
+	}
+
+	if cap(b) < p.blen || cap(b) > p.blen*2 {
+		// DO NOT reuse buffers with insufficient capacity.
+		// This could cause panics when resizing to p.blen.
+
+		// DO NOT reuse buffers with excessive capacity.
+		// This could cause memory leaks.
+		return
+	}
+
```

Comment:
This is just a whole bunch of defensive programming. Now that the `bufPool`s are stretching out in usage, it became clear that we needed this to be way safer than it was. The old version could possibly panic, required us to read in and put back slices in special cases.

Now, as long as we put the slice back in with the same start, and capacity, we’re golden. Even if we truncated its length.